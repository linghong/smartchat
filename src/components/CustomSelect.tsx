import React, { useState } from 'react';

interface Option {
  value: string;
  label: string;
}

interface CustomSelectProps {
  id: string;
  options: Option[];
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
}

const CustomSelect: React.FC<CustomSelectProps> = ({
  id,
  options,
  value,
  onChange,
  placeholder = 'Select an option'
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleToggle = () => setIsOpen(!isOpen);

  const handleSelect = (option: Option) => {
    onChange(option.value);
    setIsOpen(false);
  };

  const selectedOption = options.find(option => option.value === value);

  return (
    <div id={id} className="relative">
      <div
        className="select-trigger w-full p-2 border border-gray-300 rounded-md cursor-pointer bg-white text-left"
        onClick={handleToggle}
      >
        <span className={selectedOption ? 'text-gray-900' : 'text-gray-500'}>
          {selectedOption ? selectedOption.label : placeholder}
        </span>
      </div>
      {isOpen && (
        <ul className="absolute z-10 w-full mt-1 bg-white border border-gray-300 rounded-md shadow-lg max-h-60 overflow-auto">
          {options.map(option => (
            <li
              key={option.value}
              onClick={() => handleSelect(option)}
              className={`px-4 py-2 cursor-pointer hover:bg-gray-100 ${
                option.value === value
                  ? 'bg-slate-200 text-slate-900'
                  : 'text-gray-900'
              }`}
            >
              {option.label}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default CustomSelect;
