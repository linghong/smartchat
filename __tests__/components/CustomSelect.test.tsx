import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import CustomSelect from '@/src/components/CustomSelect';
const mockOptions = [
  { value: 'option1', label: 'Option 1' },
  { value: 'option2', label: 'Option 2' },
  { value: 'option3', label: 'Option 3' }
];
describe('CustomSelect', () => {
  const mockOnChange = jest.fn();
  beforeEach(() => {
    mockOnChange.mockClear();
  });
  it('renders with placeholder when no value is selected', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value=""
        onChange={mockOnChange}
        placeholder="Select an option"
      />
    );
    expect(screen.getByText('Select an option')).toBeInTheDocument();
  });
  it('renders with selected value when provided', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value="option2"
        onChange={mockOnChange}
      />
    );
    expect(screen.getByText('Option 2')).toBeInTheDocument();
  });
  it('opens dropdown when clicked', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value=""
        onChange={mockOnChange}
      />
    );
    fireEvent.click(screen.getByText('Select an option'));
    mockOptions.forEach(option => {
      expect(screen.getByText(option.label)).toBeInTheDocument();
    });
  });
  it('calls onChange with correct value when option is selected', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value=""
        onChange={mockOnChange}
      />
    );
    fireEvent.click(screen.getByText('Select an option'));
    fireEvent.click(screen.getByText('Option 2'));
    expect(mockOnChange).toHaveBeenCalledWith('option2');
  });
  it('closes dropdown after selecting an option', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value=""
        onChange={mockOnChange}
      />
    );
    fireEvent.click(screen.getByText('Select an option'));
    fireEvent.click(screen.getByText('Option 2'));
    expect(screen.queryByText('Option 1')).not.toBeInTheDocument();
    expect(screen.queryByText('Option 3')).not.toBeInTheDocument();
  });
  it('applies correct styles to selected option', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value="option2"
        onChange={mockOnChange}
      />
    );
    //open the dropdown
    fireEvent.click(screen.getByText('Option 2'));

    const dropdown = screen.getByRole('list');
    const selectedOption = within(dropdown).getByText('Option 2');
    expect(selectedOption).toHaveClass('bg-slate-200');
    expect(selectedOption).toHaveClass('text-slate-900');
  });
  it('applies hover styles to options', () => {
    render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value=""
        onChange={mockOnChange}
      />
    );
    fireEvent.click(screen.getByText('Select an option'));

    const option = screen.getByText('Option 1');
    fireEvent.mouseEnter(option);

    expect(option).toHaveClass('hover:bg-gray-100');
  });

  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(
      <CustomSelect
        id="test-select"
        options={mockOptions}
        value="option2"
        onChange={mockOnChange}
        placeholder="Select an option"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
