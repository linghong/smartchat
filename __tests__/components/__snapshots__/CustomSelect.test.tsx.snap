// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSelect renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative"
    id="test-select"
  >
    <div
      class="select-trigger w-full p-2 border border-gray-300 rounded-md cursor-pointer bg-white text-left"
    >
      <span
        class="text-gray-900"
      >
        Option 2
      </span>
    </div>
  </div>
</DocumentFragment>
`;
